### PR TITLE
[FEAT] Adding Ped Editor

### PR DESCRIPTION
**Pull request by aoqia194**
This is not a request to pull into the base repository (unless you want to).
If you do decide to, consider only pulling the changes to the code, not the README.md changes.

I have put in 2 features (Ped Editor and No Crosshair) for my own benefit, but I wanted to put them here in-case anyone wanted to use them.

 - Ped Editor just lets you change the drawable and texture ids for clothing and props (Watches and Bracelets don't work in Story Mode from my testing)
 - No Crosshair just removes the reticle every frame (opposite of Force Crosshair) in-case you wanted to use a custom crosshair that overlays onto the game.

If you think you could request changes to my code to improve it, or if you have a better idea for where I should put this, I am happy to listen.

### CODE REVIEW

**Comment by xiaoxiao921** [comment]
Looks good to me but you'll have to remove these readme change if you want the pr to be merged

File: src/backend/looped/weapons/no_crosshair.cpp [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#include "backend/looped/looped.hpp"
+#include "natives.hpp"
+
+namespace big
+{
+	void looped::weapons_no_crosshair()
+	{
+		if (g->weapons.no_crosshair) {
+			HUD::HIDE_HUD_COMPONENT_THIS_FRAME(14 /*RETICLE*/);
```
Comment (tupoy-ya) [inline src/backend/looped/weapons/no_crosshair.cpp]: Can't you already do that with #553?

https://github.com/YimMenu/YimMenu/blob/1087146e5670f6710a84cc841586a18f6db8055c/src/backend/looped/self/hud.cpp#L45-L47

Comment (aoqia194) [reply src/backend/looped/weapons/no_crosshair.cpp]: All I did was took the functionality of src/backend/looped/weapons/force_crosshair.cpp and copy pasted it to hide the component instead. I might switch to what you suggested though.

**Comment by aoqia194** [comment]
Also do know that I didn't implement any config saving for the no crosshair as when I did it was incompatible with older configs. (Don't know if that's a problem or not but I considered it anyways)
And I didn't do it for the ped editor either because if you wanted to save your ped's clothing and props you can just use the outfit editor in clothing shop/apartment.

File: src/backend/looped/weapons/no_crosshair.cpp (hunk 1 above)
Comment (Yimura) [reply src/backend/looped/weapons/no_crosshair.cpp]: Nah he's just saying that the functionality already exists thus it's no longer required.
So, you can just remove the code to hide the crosshair.
According to the code it should be possible to hide the RETICLE.

Comment (aoqia194) [reply src/backend/looped/weapons/no_crosshair.cpp]: I see what you mean now. I will remove that rn

**Review by aoqia194 — COMMENTED** [review]
~snip~
I am just going to test it before any more things happen 😵‍💫

**Comment by aoqia194** [comment]
Here is a preview of the imgui layout in the menu
I am testing it both in SP and MP and it works fine on my end.

Here is some IDs https://wiki.rage.mp/index.php?title=Clothes
Please review my code and test before merging (for my anxiousness)

**Comment by xiaoxiao921** [comment]
The CI is failing to compile the menu, please fix the compilation errors

**Comment by aoqia194** [comment]
> The CI is failing to compile the menu, please fix the compilation errors

My error, I forgot to push my changes on my end :/

**Comment by Yimura** [comment]
Please fix the following issues to the code:
image

Next time try cloning the code locally, that way you can compile the code to test if you didn't mess something up.
You can also commit everything at once when checking out locally.

**Comment by xiaoxiao921** [comment]
Also, is there any way that we show some names instead of numbers? People will need to use that webpage you linked which can be annoying

**Comment by aoqia194** [comment]
> Please fix the following issues to the code: ~snip~
> 
> Next time try cloning the code locally, that way you can compile the code to test if you didn't mess something up. You can also commit everything at once when checking out locally.

Yes, sorry I am very forgetful and always feel like there is "just one more thing" right after I commit my changes but good suggestion

**Comment by aoqia194** [comment]
> Also, is there any way that we show some names instead of numbers? People will need to use that webpage you linked which can be annoying

Good suggestion I will implement that right now
I did use `PED::GET_PED_PROOP_DRAWABLE_VARATIONS` to get the max drawables and textures and clamp the user input ids to that but it kept giving me annoying access violation so I deleted it to focus more on the main code

**Comment by aoqia194** [comment]
@xiaoxiao921 Also, I am confused on what you mean by names?
You mean like having a list of items with the name of the items? (Would that suffice)

**Comment by tupoy-ya** [comment]
I think can use `GET_HASH_NAME_FOR_COMPONENT` and `GET_HASH_NAME_FOR_PROP` to get a hash name and then use
`GET_SHOP_PED_COMPONENT` to get the actual name.

Didn't test it just looked at leaked scripts.

EDIT: It only works online.

**Comment by aoqia194** [comment]
> I think can use `GET_HASH_NAME_FOR_COMPONENT` and `GET_HASH_NAME_FOR_PROP` to get a hash name and then use `GET_SHOP_PED_COMPONENT` to get the actual name.
> 
> Didn't test it just looked at leaked scripts.

Thank you so much that actually helped.
~snip~ FIXED MY ISSUE

**Comment by tupoy-ya** [comment]
Here is my test code for getting this label
```cpp
struct PedComponent
{
	int64_t m_lockHash;
    	int64_t m_nameHash;
    	int64_t m_locate;
    	int64_t m_drawableIndex;
    	int64_t m_textureIndex;
    	int64_t m_cost;
    	int64_t m_eComponentType;
    	int64_t m_eShopEnum;
    	int64_t m_eCharacter;
    	char m_textLabel[64];
};

components::button("Test component", [ped_to_edit]
{
	PedComponent component;
	Hash hash = FILES::GET_HASH_NAME_FOR_COMPONENT(ped_to_edit, g->outfit_editor.componentId, g->outfit_editor.drawableId, g->outfit_editor.textureId);
	FILES::GET_SHOP_PED_COMPONENT(hash, (Any*)&component);
	LOG(G3LOG_DEBUG) << HUD::GET_FILENAME_FOR_AUDIO_CONVERSATION(component.m_textLabel);
});
```

**Comment by aoqia194** [comment]
Getting the editor to work with names will be quite a tough challenge for me. I can do everything fine but for whatever reason, using the `PED`::`GET_NUMBER_OF_PED_PROP_XXX_VARIATIONS` crashes the game. I've tried calling the native once a frame, once every time a component updates, etc and it always ends up crashing (Even if it's after I call it a couple times). I've tried caching, same results.
It's driving me crazy honestly. Also for some reason I couldn't seem to get the names correctly (they were always returning "NULL" using the `HUD`::`GET_FILENAME_FOR_AUDIO_CONVERSATION` native)

If anyone can seem to get the number variations natives working (or something similar that has the same result) then I will be able to implement it.

**Comment by tupoy-ya** [comment]
> Getting the editor to work with names will be quite a tough challenge for me. I can do everything fine but for whatever reason, using the `PED`::`GET_NUMBER_OF_PED_PROP_XXX_VARIATIONS` crashes the game. I've tried calling the native once a frame, once every time a component updates, etc and it always ends up crashing.

Do you put natives in fiber pool?
```cpp
if (ImGui::InputInt("Drawable Id", &g->outfit_editor.drawableId)) {
	g_fiber_pool->queue_job( [ped_to_edit] {
		PED::SET_PED_COMPONENT_VARIATION(ped_to_edit, g->outfit_editor.componentId, g->outfit_editor.drawableId, g->outfit_editor.textureId, g->outfit_editor.paletteId);
	});
}
```

> Also for some reason I couldn't seem to get the names correctly (they were always returning "NULL" using the HUD::GET_FILENAME_FOR_AUDIO_CONVERSATION native)

Had the same problem `FILES::GET_HASH_NAME_FOR_COMPONENT` only works on online peds for some reason.
I removed some DEBUG logs before sending my test code i might have changed something else and forgot about it.

**Comment by aoqia194** [comment]
Oh my. I didn't even know the fiber pool was a thing. Do you know the
reason for using fiber pool over normal function call (trying to learn what
it is)

But I will fix this issues just not now. I have to sleep then work then I
will come home after and fix the stuff.

**Comment by xiaoxiao921** [comment]
You can only call natives from there, for various reasons, one of them being the fiber pool execute the functions you pass it from a Thread that has the required context, we call them Gta Thread

**Comment by Yimura** [comment]
Personally, the UI seems fine as-is and as you've never worked with a fiber pool it might be hard for you to communicate between the fiber pool and the UI (frontend).

Unless you really want to add it in this same PR, I'm fine with merging this without the changes requested by @tupoy-ya.

**Comment by aoqia194** [comment]
I'll see if I can get thr names working this afternoon but if I can't ill
make a small change to the component names and then I'll let you know to
merge it

**Comment by aoqia194** [comment]
It's not far off being done, I just have to fix a crash or two. I used the fiber pool aswell.

**Comment by aoqia194** [comment]
Bumping this because I need help with an issue (please read review comments and if you have a solution or any tips for my code please let me know)

File: src/util/ped_editor.hpp [hunk 2]
```diff
@@ -1,170 +1,287 @@
...
-		return component_name;
-	}
-}
+#pragma once
+#include "natives.hpp"
+#include "pointers.hpp"
+#include "script.hpp"
+
+template <typename t, size_t szof>
+constexpr int countof(t (&array)[szof]) { return szof; }
```
Comment (SpaghettDev) [inline src/util/ped_editor.hpp]: just use ImGui's `IM_ARRAYSIZE`

Comment (Yimura) [reply src/util/ped_editor.hpp]: Just a simple `sizeof()` should work there as well, the compiler should be smart enough to substitute that to the correct size.

**Comment by maybegreat48** [comment]
You are crashing because they way you are using the fiber pool is completely wrong. Fiber pools are queued to be run in the next game tick so you can't expect the value to be populated when you return from the function. A solution to this is to use static variables and some kind of `ready` bool to show the UI after everything you need is populated into the static variables

I see you are taking references to stack variables in the fiber callback which is a very bad idea and is probably what's causing your crash

**Pull request closed without merging**